dace/transformation/dataflow/vectorization.py

```diff
@@ -18,100 +143,341 @@ class Vectorization(transformation.SingleStateTransformation):
 
-        return found
+                if ranges_list[-1][2] != 1:
+                    return False
+
+            # Check all edges
+            for m in maps_to_vectorize:
+
+                correct_state = get_innermost_state_for_node(sdfg, m)
+
+                edges = correct_state.all_edges(m)
+
+                map_subset = m.map.params
+
+                for e in edges:
+
+                    if e is None or e.data is None or e.data.subset is None:
+                        continue
+
+                    edge_subset = [a_tuple[0] for a_tuple in list(e.data.subset)]
+
+                    if isinstance(edge_subset[-1], symbol) and str(edge_subset[-1]) != map_subset[-1]:
+                        return False
+
+            # Not possible to handle interstate edges
+
+            for e, _ in sdfg.all_edges_recursive():
+
+                if isinstance(e.data, InterstateEdge):
+
+                    if e.data.assignments != {} or e.data.condition.as_string != "1":
+                        return False
+
+            self.map_entry = old_map_entry
+            self._level = 0
+
+            # Run the vector inference algorithm to check if vectorization is feasible
+        try:
+            vector_inference.infer_vectors(sdfg,
+                                           state,
+                                           map_entry,
+                                           self.vector_len,
+                                           flags=vector_inference.VectorInferenceFlags.Allow_Stride,
+                                           strided_map=self.strided_map
+                                           or self.target == dtypes.ScheduleType.FPGA_Device,
+                                           apply=False)
+        except vector_inference.VectorInferenceException as ex:
+            return False
+
+        return True
 
     def apply(self, graph: SDFGState, sdfg: SDFG):
         map_entry = self.map_entry
         tasklet: nodes.Tasklet = graph.successors(map_entry)[0]
         param = symbolic.pystr_to_symbolic(map_entry.map.params[-1])
 
-        # Create new vector size.
+        # To support recursivity in the FPGA case, see below
+        if self._level == 0:
+            state = graph
+        else:
+            state = get_innermost_state_for_node(sdfg, map_entry)
+
+        if self.target == dtypes.ScheduleType.FPGA_Device and self._level == 0:
+
+            maps_to_vectorize, data_descriptors_to_vectorize = collect_maps_to_vectorize(sdfg, state, map_entry)
+
+            old_map_entry = self.map_entry
+            self._level = 1  # To prevent infinte loop
+
+            for m in maps_to_vectorize:
+                self.map_entry = m
+
+                self.apply(get_innermost_state_for_node(sdfg, m), get_innermost_sdfg_for_node(sdfg, m))
+
+            # Change the subset in the post state that copies the data back to the host
+            if len(sdfg.states()) < 3 or not sdfg.states()[-1].name.startswith('post_'):
```

Comment:
We cannot specialize this to _only_ work for when the FPGA transformation has been applied. What if I implemented this manually? I could name it anything I want.